src/keygen/Avvm.hs

```diff
@@ -75,24 +108,36 @@ genGenesis avvm genCerts holder = GenesisData
     balances :: HashMap Address (Coin, TxOutDistribution)
     balances = HM.fromListWith sumOutcomes $ do
         AvvmEntry{..} <- utxo avvm
-        let pk = case decodeUrl address of
-                Right x -> RedeemPublicKey (Ed.PublicKey x)
-                Left _  -> error ("couldn't decode address " <> address)
-        let addr = makeRedeemAddress pk
-            adaCoin = unsafeIntegerToCoin $ coinAmount coin
+        let addr = makeRedeemAddress aePublicKey
+            adaCoin = unsafeIntegerToCoin aeCoin
         return (addr, (adaCoin, distr adaCoin))
 
+-- | Applies blacklist to avvm utxo, produces warnings and stats about
+-- how much was deleted.
+applyBlacklisted :: Maybe FilePath -> AvvmData -> IO AvvmData
+applyBlacklisted Nothing r = r <$ putText "Blacklisting: file not specified, skipping"
+applyBlacklisted (Just blacklistPath) AvvmData{..} = do
+    addrTexts <- T.lines <$> TIO.readFile blacklistPath
+    blacklisted <- mapM fromAvvmPk addrTexts
+    let filteredBad = filter ((`elem` blacklisted) . aePublicKey) utxo
+    let filtered = utxo \\ filteredBad
+    putText $
+        "Removing " <> show (length filtered) <> " entries from utxo (out of " <>
+        show (length blacklisted) <> " total entries in the blacklist)"
+    pure $ AvvmData filtered
+
 getHolderId :: Maybe FilePath -> IO StakeholderId
-getHolderId holderPath = case holderPath of
-    Just fileName -> do
-        mSk <- view usPrimKey <$> readUserSecret fileName
-        sk <- maybe (fail "No secret key is found in file")
-              pure mSk
-        pure $ addressHash $ toPublic sk
-    Nothing -> do
-        putText "USING RANDOM STAKEHOLDER ID."
-        putText "NOT FOR PRODUCTION USAGE, ONLY FOR TESTING"
-        putText "IF YOU INTEND TO GENERATE GENESIS.BIN FOR PRODUCTION, \
-                \STOP RIGHT HERE AND USE `--fileholder <path to secret>` OPTION. \
-                \THIS IS SERIOUS."
-        addressHash . fst <$> keyGen
+getHolderId (Just fileName) = do
+    mSk <- view usPrimKey <$> readUserSecret fileName
+    let sk = fromMaybe (error "No secret key is found in file") mSk
+    pure $ addressHash $ toPublic sk
+getHolderId Nothing = do
+    skPath <- ("redeemingHolderKey" <>) . show <$> randomRIO (0,100000::Int)
```

Comment:
I think there should be nice way to convert `Int` to `Text` w/o `show` and further conversion from `String` to `Text`